File: backend/app/schemas/diagnosis.py
"""Request / response bodies for diagnoses."""

from pydantic import BaseModel, ConfigDict, Field


class PredictionOut(BaseModel):
    disease_key: str
    confidence: float = Field(ge=0.0, le=1.0)


class HeatmapRegion(BaseModel):
    """One highlighted ellipse, in normalised 0..1 image coordinates.

    Normalised so the app can draw the overlay at whatever size it renders the
    photo, without knowing the resolution the model saw.
    """

    cx: float = Field(ge=0.0, le=1.0)
    cy: float = Field(ge=0.0, le=1.0)
    rx: float = Field(gt=0.0, le=1.0)
    ry: float = Field(gt=0.0, le=1.0)
    weight: float = Field(default=1.0, ge=0.0, le=1.0)


class Heatmap(BaseModel):
    regions: list[HeatmapRegion]
    """Share of the leaf the model flagged, 0..1 — shown as "Vùng nghi ngờ"."""
    affected_ratio: float | None = Field(default=None, ge=0.0, le=1.0)


class PredictResponse(BaseModel):
    """Top-3 predictions, best first.

    No database row is created — see the note at the top of routers/diagnoses.py.
    """

    predictions: list[PredictionOut]
    heatmap: Heatmap | None = None
    model_version: str
    image_path: str
    analysed_at: int


class DiagnosisOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    plot_id: str
    disease_key: str
    disease_name: str
    severity: str
    confidence: float
    affected_ratio: float | None
    image_path: str | None
    model_version: str | None
    explanation: str | None
    top3_json: str | None
    heatmap_json: str | None
    queued: bool
    diagnosed_at: int
    owner_id: str
    created_at: int
    updated_at: int

File: mobile/src/db/models/Diagnosis.ts
import {Model} from '@nozbe/watermelondb';
import {date, field, readonly, relation, text} from '@nozbe/watermelondb/decorators';
import type {Associations} from '@nozbe/watermelondb/Model';

import type Plot from './Plot';

export type SeverityKey = 'none' | 'mild' | 'moderate' | 'severe';

export default class Diagnosis extends Model {
  static table = 'diagnoses';

  static associations: Associations = {
    plots: {type: 'belongs_to', key: 'plot_id'},
  };

  @text('plot_id') plotId!: string;
  @text('disease_key') diseaseKey!: string;
  @text('disease_name') diseaseName!: string;
  @text('severity') severity!: SeverityKey;
  @field('confidence') confidence!: number;
  @field('affected_ratio') affectedRatio!: number | null;
  @text('image_path') imagePath!: string | null;
  @text('model_version') modelVersion!: string | null;
  @text('explanation') explanation!: string | null;
  /** JSON: the three predictions the model returned, best first. */
  @text('top3_json') top3Json!: string | null;
  /** JSON heatmap, or null when the model produced none — never faked. */
  @text('heatmap_json') heatmapJson!: string | null;
  /** True while the photo is waiting for a network connection (Điều 3). */
  @field('queued') queued!: boolean;
  @field('diagnosed_at') diagnosedAt!: number;
  @text('owner_id') ownerId!: string;
  @text('updated_by') updatedBy!: string | null;
  @readonly @date('created_at') createdAt!: Date;
  @readonly @date('updated_at') updatedAt!: Date;

  @relation('plots', 'plot_id') plot!: Plot;
}
